### PR TITLE
[FLINK-24708][planner] Fix wrong results of the IN operator

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Fix the wrong results of the IN opreator. The reason behind is it uses the type of the first element in the SEARCH argument as the type of the other elements. It may change the value of the other elements. For example, it may rightPad
the char element if the common type has larger size. Therefore, it recommends to use `RelDataTypeFactory` to calculate the least restrictive type for all types.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala [hunk 1]
```diff
@@ -184,6 +188,52 @@ class ConvertToNotInOrInRule
       case _ => rexNodes.size >= THRESHOLD
     }
   }
+
+
+  /**
+   * Convert the conditions into the [[IN]] and fix the bug in [[RexBuilder#makeIn]].
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala:194]: Is there any Calcite issue tracking this bug？

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRuleTest.scala [hunk 2]
```diff
@@ -169,4 +169,32 @@ class ConvertToNotInOrInRuleTest extends TableTestBase {
...
+  }
+
+  @Test
+  def testConvertToSearch2(): Unit = {
+    util.verifyRelPlan(
+      "SELECT * FROM MyTable WHERE " +
+        "e = 'a' or e = 'b' or e = 'c' or e = 'd' or e = 'e' or e = 'f' or e = NULL or e = " +
+        "'HELLO " +
+        "WORLD!'"
+    )
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRuleTest.scala:198]: please add some tests to cover other types (such as, int, bigint, double)

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/CalcITCase.scala [hunk 3]
```diff
@@ -1419,6 +1419,76 @@ class CalcITCase extends BatchTestBase {
...
+        |
+        |""".stripMargin
+    checkResult(
+      sql,
+      Seq(row("HC809"), row("H389N     "))
+    )
+  }
+
+  @Test
+  def testSearchWithNull(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/CalcITCase.scala:1459]: please add a test that the first element is null

**Review by godfreyhe — COMMENTED** [review]
Thanks for the fix @fsk119 , I left a few comments

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala:194]: https://issues.apache.org/jira/browse/CALCITE-4888

**Review by godfreyhe — APPROVED** [review]
LGTM

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala [hunk 4]
```diff
@@ -184,6 +188,54 @@ class ConvertToNotInOrInRule
       case _ => rexNodes.size >= THRESHOLD
     }
   }
+
```
Comment (beyond1920) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala:191]: Could we use the following method to fix?
image

**Review by beyond1920 — COMMENTED** [review]
@fsk119 @godfreyhe I think it's a bug in Calcite.
makeIn does not take `FlinkTYpeSystem.shouldConvertRaggedUnionTypesToVarying` into consideration. When this is true, the behavior should not  padding char. Please see https://issues.apache.org/jira/browse/CALCITE-4590 and https://issues.apache.org/jira/browse/CALCITE-2321.
In flink, `FlinkTypeSystem ` overwrite `FlinkTYpeSystem.shouldConvertRaggedUnionTypesToVarying`, it always return true.
I guess we could update the `RexBuilder.makeIn` method to use `getTypeFactory().leastRestrictive(rangesTypes)` instead of `range0Type`.
WDYT?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala (hunk 4 above)
Comment (beyond1920) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRule.scala:191]: Sorry, I checked the scalar code in currently `ConvertToNotInOrInRule.scala`, they are same with the above java code.
Could we move this method to `FlinkRexBuilder`, and overwrite the makeIn method of parent class?
Then we could avoid this bug appears in other places.

**Review by beyond1920 — APPROVED** [review]

**Review by beyond1920 — APPROVED** [review]

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRuleTest.scala [hunk 5]
```diff
@@ -169,4 +169,40 @@ class ConvertToNotInOrInRuleTest extends TableTestBase {
...
+        |'HC809',
+        |'DH-9908N-AEP',
+        |'DH-9908N'
+        |)
+        |""".stripMargin
+    )
+  }
+
+  @Test
+  def testConvertToSearchString2(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/ConvertToNotInOrInRuleTest.scala:192]: rename to `testConvertToSearchStringWithNull` ?

**Review by godfreyhe — APPROVED** [review]
thanks for the update, LGTM, I will merge it

**Pull request closed without merging**